README
NAME
    WWW::OpenResty - Client-side library for OpenResty servers

VERSION
    This document describes "WWW::OpenResty" 0.09 released on September 25,
    2008.

SYNOPSIS
        use WWW::OpenResty;

        my $resty = WWW::OpenResty->new(
            { server => 'http://resty.eeeeworks.org' }
        );

        # returns an HTTP::Response object
        my $res = $resty->get('/=/version');
        if ($res->is_success) {
            print $res->content;
        } else {
            die $res->status_line;
        }

        $res = $resty->get(
            '/=/model/Post/~/~',
            { user => 'agentzh.Public', offset => 3, limit => 10 }
        );

        $resty->login($user, $password);
        $resty->delete('/=/model');  # delete all the models
        $resty->delete('/=/model/Foo');  # delete the Foo model
        $res = $resty->get('/=/role');  # get the role list

        # create model Post
        $res = $resty->post(
            '/=/model/Post.json',
            '{
                description: "Blog post",
                columns: [
                    { name: "title", label: "Post title" },
                    { name: "content", label: "Post content" },
                    { name: "author", label: "Post author" },
                    { name: "created", default: ["now()"],
                    type: "timestamp(0) with time zone",
                    label: "Post creation time" },
                    { name: "comments", label: "Number of comments",
                    default: 0 }
                ]
            }'
        );

        # modify the label for the model column "title":
        $res = $resty->put('/=/model/Post/title', '{ label: "blah!" }');

DESCRIPTION
    "WWW::OpenResty" wraps an LWP::UserAgent object and serves as a client
    for OpenResty servers.

    This module is still under ative development and we definitely need more
    POD and tests. We're just following the "release early, release often"
    guideline. Please check back often :P

    But most of the time, you just need its subclass WWW::OpenResty::Simple
    which provides a more friendly interface and with automatic error
    checking support.

METHODS
    "$res = $obj->login($user, $password)"
    "$res = $obj->get($url)"
    "$res = $obj->get($url, $url_params)"
    "$res = $obj->post($url, $content)"
    "$res = $obj->post($url, $url_params, $content)"
    "$res = $obj->put($url, $content)"
    "$res = $obj->put($url, $url_params, $content)"
    "$res = $obj->delete($url)"
    "$res = $obj->delete($url, $url_params)"

SOURCE CONTROL
    For the very latest version of this module, check out the source from
    the SVN repos below:

    <http://svn.openfoundry.org/wwwopenresty>

    There is anonymous access to all. If you'd like a commit bit, please let
    us know. :)

BUGS
    Please report bugs or send wish-list to
    <http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-OpenResty>.

SEE ALSO
    OpenResty, HTTP::Response.

AUTHOR
    agentzh "<agentzh at yahoo.cn>"

COPYRIGHT
    Copyright (c) 2008 by Yahoo! China EEEE Works, Alibaba Inc.

License
    The "MIT" License

    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
Change Log for WWW::OpenResty

0.09 Sep 25 2008
* Fixed the /=/version test since the Version API changed in a recent OpenResty release.

0.08 Sep 2 2008
* Fixed the url params handling logic in request.

0.07 Sep 2 2008
* Use _use_cookie and _user options rather than the deprecated use_cookie and user options.
* Now the login method allow signing in without a password.

0.06 Apr 21
* Now use MD5 to preprocess the password for login (working with OpenResty server 0.2.x).

0.05 Apr 4 2008
* Added the retries option to WWW::OpenResty::Simple's constructor.

0.04 Mar 29 2008
* Fixed a bug in the has_model and has_view methods.

0.03 Mar 28 2008
* Added methods has_model and has_view to WWW::OpenResty::Simple.
* Using croak now for better diagnostics.
* Now we use JSON::XS instead of JSON::Syck.

0.02 Mar 19 2008
* Changed the param order of the post and put methods of
  both WWW::OpenResty and WWW::OpenResty::Simple.
* Added POD to WWW::OpenResty::Simple.

0.01 Mar 3 2008
* initial CPAN release



lib/WWW/OpenResty/Simple.pm
package WWW::OpenResty::Simple;

use strict;
use warnings;

use Carp 'croak';
use JSON::XS ();
use base 'WWW::OpenResty';
use Params::Util qw( _HASH );

our $VERSION = '0.09';
our $json_xs = JSON::XS->new->utf8->allow_nonref;

sub request {
    my $self = shift;
    my $data = $_[0];
    my $meth = $_[1];
    my $url = $_[2];
    if ($data && ref $data) {
        $_[0] = $json_xs->encode($data);
    }
    my $retries = $self->{retries} || 0;
    my $ignore_dup_error = $self->{ignore_dup_error};
    my $i = 0;
    while (1) {
        my $res = $self->SUPER::request(@_);
        if ($res->is_success) {
            my $json = $res->content;
            #$json =~ s/\n+$//gs;
            my $data = $json_xs->decode($json);
            if (_HASH($data) && defined $data->{success} && $data->{success} == 0) {
                if ($ignore_dup_error &&
                    $data->{error} =~ /duplicate key (?:value )?violates unique constraint/) {
                    #print "Ignored
                    return $data;
                }
                if ($i >= $retries) {
                    croak "$meth $url: $json";
                } else {
                    warn "Retrying...\n";
                }
            } else {
                return $data;
            }
        } else {
            my $status_line = $res->status_line;
            if ($i >= $retries) {
                croak "$meth $url: $status_line";
            } else {
                warn "Retrying...\n";
            }
        }
    } continue { $i++ }
}

sub has_model {
    my ($self, $model) = @_;
    my $res;
    eval {
        $res = $self->get("/=/model/$model");
    };
    if ($@) {
        if ($@ =~ /Model .*? not found/i) {
            return undef;
        }
        die $@;
    }
    return _HASH($res) && $res->{name} eq $model;
}

sub has_view {
    my ($self, $view) = @_;
    my $res;
    eval {
        $res = $self->get("/=/view/$view");
    };
    if ($@) {
        if ($@ =~ /View .*? not found/i) {
            return undef;
        }
        die $@;
    }
    return _HASH($res) && $res->{name} eq $view;
}

1;
__END__

=head1 NAME

WWW::OpenResty::Simple - A simple wrapper around WWW::OpenResty

=head1 VERSION

This document describes C<WWW::OpenResty::Simple> 0.09 released on September 25,
2008.

=head1 SYNOPSIS

    use WWW::OpenResty::Simple;

    my $resty = WWW::OpenResty::Simple->new(
        {
            server => 'http://resty.eeeeworks.org',
            retries => 1 # retry once when failing (default to 0),
        }
    );

    my $res;
    eval {
        $res = $resty->get('/=/version');
    };
    if ($@) { die $@ }
    print Dumper($res);
    # Got:
    # $VAR1 = 'OpenResty 0.1.0 (revision 973) with the ...'

    # The following requires privileges...
    $resty->login('account.Role', 'password');
    $res = $resty->post(
        '/=/model/Comments/~/~',
        {
            author:"agentzh",
            title:"Great post!",
            body:"clapping..."
        }
    );

    # Update the comment with ID 3:
    $res = $resty->put(
        '/=/model/Comments/id/3',
        { title: "New title" }
    );

    # Remove the comment with the ID 3:
    $res = $resty->delete('/=/model/Comments/id/3');

=head1 DESCRIPTION

This class inherits from L<WWW::OpenResty>, which does automatic JSON
serialization for input and output data. In addition, it croaks on
errors returned from the OpenResty server.

=head1 METHODS

This class exposes all the methods of its parent class, L<WWW:OpenResty>,
like C<login>, C<get>, C<post>, C<put>, and C<delete>, as well as its own:
C<has_view> and C<has_model>.

Unlike the methods of L<WWW::OpenResty> returning a raw L<HTTP::Response>
object, the methods in this class always return the Perl structure
directly. That is, it automatically load the JSON literals sent by
the server if no error occurs; it throws an exception out otherwise.

Also, the C<$content> parameter in the C<post> and C<put> methods
are plain Perl data structures rather than string literals. This class will
automatically serialize the data structure into JSON before sending
out via the HTTP protocol.

=head1 METHODS

=over

=item C<< $bool = $resty->has_view($view_name) >>

Check if the view given exists.

=item C<< $bool = $resty->has_model($model_name) >>

Check if the model given exists.

=back

=head1 AUTHOR

agentzh C<< <agentzh at yahoo.cn> >>

=head1 COPYRIGHT

Copyright (c) 2008 by Yahoo! China EEEE Works, Alibaba Inc.

=head1 License

The "MIT" License

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



lib/WWW/OpenResty.pm
package WWW::OpenResty;

use strict;
use warnings;

#use Smart::Comments;
use Carp qw(croak);
use Params::Util qw( _HASH0 );
use LWP::UserAgent;
use Data::Dumper;
use Digest::MD5 qw(md5_hex);

our $VERSION = '0.09';

sub new {
    ### @_
    my $class = ref $_[0] ? ref shift : shift;
    my $params = _HASH0(shift @_) or croak "Invalid params";
    ### $params
    my $server = delete $params->{server} or
        croak "No server specified.";
    if ($server !~ m{^\w+://}) {
        $server = "http://$server";
    }
    my $ignore_dup_error = delete $params->{ignore_dup_error};
    my $timer = delete $params->{timer};
    my $retries = delete $params->{retries};
    my $ua = LWP::UserAgent->new;
    $ua->cookie_jar({ file => "cookies.txt" });
    bless {
        server => $server,
        ua => $ua,
        timer => $timer,
        retries => $retries,
        ignore_dup_error => $ignore_dup_error,
    }, $class;
}

sub content_type {
    $_[0]->{content_type} = $_[1];
}

sub login {
    my ($self, $user, $password) = @_;
    if ($password) {
        $password = md5_hex($password);
        return $self->get("/=/login/$user/$password?_use_cookie=1");
    }
    $self->{_user} = $user;
}

sub get {
    my $self = shift;
    $self->request(undef, 'GET', @_);
}

sub post {
    my $self = shift;
    my $content = pop;
    $self->request($content, 'POST', @_);
}

sub put {
    my $self = shift;
    my $content = pop;
    $self->request($content, 'PUT', @_);
}

sub delete {
    my $self = shift;
    $self->request(undef, 'DELETE', @_);
}

sub request {
    my ($self, $content, $method, $url, $params) = @_;
    !defined $params or _HASH0($params) or
        croak "Params must be a hash: ", Dumper($params);
    !ref $url or croak "URL is of the wrong type: ", Dumper($url);
    $params ||= {};
    if ($self->{_user}) {
        $params->{_user} ||= $self->{_user};
    }
    my @params;
    while (my ($key, $val) = each %$params) {
        push @params, "$key=$val";
    }
    if ($url =~ /\?$/) {
        $url .= join '&', @params;
    } elsif ($url =~ /\?/) {
        $url .= "&" . join '&', @params;
    } else {
        $url .= "?" . join '&', @params;
    }
    my $type = $self->{content_type};
    $type ||= 'text/plain';
    if ($url !~ /^http:\/\//) {
        $url = $self->{server} . $url;
    }
    my $req = HTTP::Request->new($method);
    $req->header('Content-Type' => $type);
    $req->header('Accept', '*/*');
    $req->url($url);
    if ($content) {
        if ($method eq 'GET' or $method eq 'HEAD') {
            croak "HTTP 1.0/1.1 $method request should not have content: $content";
        }

        $req->content($content);
    } elsif ($method eq 'POST' or $method eq 'PUT') {
        $req->header('Content-Length' => 0);
    }
    my $timer = $self->{timer};
    my $ua = $self->{ua};
    $timer->start($method) if $timer;
    my $res = $ua->request($req);
    $timer->stop($method) if $timer;
    return $res;
}


1;
__END__

=head1 NAME

WWW::OpenResty - Client-side library for OpenResty servers

=head1 VERSION

This document describes C<WWW::OpenResty> 0.09 released on September 25, 2008.

=head1 SYNOPSIS

    use WWW::OpenResty;

    my $resty = WWW::OpenResty->new(
        { server => 'http://resty.eeeeworks.org' }
    );

    # returns an HTTP::Response object
    my $res = $resty->get('/=/version');
    if ($res->is_success) {
        print $res->content;
    } else {
        die $res->status_line;
    }

    $res = $resty->get(
        '/=/model/Post/~/~',
        { user => 'agentzh.Public', offset => 3, limit => 10 }
    );

    $resty->login($user, $password);
    $resty->delete('/=/model');  # delete all the models
    $resty->delete('/=/model/Foo');  # delete the Foo model
    $res = $resty->get('/=/role');  # get the role list

    # create model Post
    $res = $resty->post(
        '/=/model/Post.json',
        '{
            description: "Blog post",
            columns: [
                { name: "title", label: "Post title" },
                { name: "content", label: "Post content" },
                { name: "author", label: "Post author" },
                { name: "created", default: ["now()"],
                type: "timestamp(0) with time zone",
                label: "Post creation time" },
                { name: "comments", label: "Number of comments",
                default: 0 }
            ]
        }'
    );

    # modify the label for the model column "title":
    $res = $resty->put('/=/model/Post/title', '{ label: "blah!" }');

=head1 DESCRIPTION

C<WWW::OpenResty> wraps an L<LWP::UserAgent> object and serves as a client
for L<OpenResty> servers.

This module is still under ative development and we definitely need more
POD and tests. We're just following the "release early, release often" guideline. Please check back often :P

But most of the time, you just need its subclass L<WWW::OpenResty::Simple>
which provides a more friendly interface and with automatic error checking support.

=head1 METHODS

=over

=item C<< $res = $obj->login($user, $password) >>

=item C<< $res = $obj->get($url) >>

=item C<< $res = $obj->get($url, $url_params) >>

=item C<< $res = $obj->post($url, $content) >>

=item C<< $res = $obj->post($url, $url_params, $content) >>

=item C<< $res = $obj->put($url, $content) >>

=item C<< $res = $obj->put($url, $url_params, $content) >>

=item C<< $res = $obj->delete($url) >>

=item C<< $res = $obj->delete($url, $url_params) >>

=back

=head1 SOURCE CONTROL

For the very latest version of this module, check out the source from
the SVN repos below:

L<http://svn.openfoundry.org/wwwopenresty>

There is anonymous access to all. If you'd like a commit bit, please let
us know. :)

=head1 BUGS

Please report bugs or send wish-list to
L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-OpenResty>.

=head1 SEE ALSO

L<OpenResty>, L<HTTP::Response>.

=head1 AUTHOR

agentzh C<< <agentzh at yahoo.cn> >>

=head1 COPYRIGHT

Copyright (c) 2008 by Yahoo! China EEEE Works, Alibaba Inc.

=head1 License

The "MIT" License

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

